File: linux/src/models/FirmwareFile.cpp
// FAME Smart Flasher - Linux Qt Port
// Copyright 2025 Fyrby Additive Manufacturing & Engineering
// SPDX-License-Identifier: Proprietary

#include "FirmwareFile.h"
#include <QFile>
#include <QDir>
#include <QFileInfo>
#include <QLocale>
#include <algorithm>

FirmwareFile::FirmwareFile(const QString& filePath, const QByteArray& data)
{
    // Check if this looks like a bootloader (starts at 0x0)
    // Bootloader typically has different characteristics than app:
    // - Both start with 0xE9 magic
    // - Bootloader is typically smaller (< 64KB)
    // - Merged binary name often contains "merged" or "factory"
    QString fileName = QFileInfo(filePath).fileName().toLower();
    bool isMergedBinary = fileName.contains("merged") ||
                          fileName.contains("factory") ||
                          fileName.contains("combined") ||
                          fileName.contains("full");

    // If the filename suggests it's a merged binary, flash at 0x0
    // Otherwise assume it's an app-only binary at 0x10000
    uint32_t offset = isMergedBinary ? 0x0000 : 0x10000;

    FirmwareImage image;
    image.filePath = filePath;
    image.data = data;
    image.offset = offset;
    m_images.push_back(image);
}

FirmwareFile::FirmwareFile(const std::vector<FirmwareImage>& images)
    : m_images(images)
{
    // Sort by offset
    std::sort(m_images.begin(), m_images.end(),
              [](const FirmwareImage& a, const FirmwareImage& b) {
                  return a.offset < b.offset;
              });
}

FirmwareFile FirmwareFile::fromPlatformIOBuild(const QString& dirPath)
{
    std::vector<FirmwareImage> images;
    QDir dir(dirPath);

    // Standard ESP32 flash offsets
    struct FileOffset {
        QString name;
        uint32_t offset;
    };

    const FileOffset fileOffsets[] = {
        {"bootloader.bin", 0x0000},
        {"partitions.bin", 0x8000},
        {"firmware.bin", 0x10000}
    };

    for (const auto& fo : fileOffsets) {
        QString filePath = dir.filePath(fo.name);
        if (QFile::exists(filePath)) {
            QFile file(filePath);
            if (file.open(QIODevice::ReadOnly)) {
                FirmwareImage image;
                image.filePath = filePath;
                image.data = file.readAll();
                image.offset = fo.offset;
                images.push_back(image);
                file.close();
            }
        }
    }

    if (images.empty()) {
        throw FirmwareLoadError(FirmwareLoadError::NoFilesFound,
                                "No firmware files found in directory");
    }

    // At minimum we need firmware.bin
    bool hasFirmware = std::any_of(images.begin(), images.end(),
                                   [](const FirmwareImage& img) {
                                       return img.offset == 0x10000;
                                   });
    if (!hasFirmware) {
        throw FirmwareLoadError(FirmwareLoadError::MissingFirmware,
                                "Missing firmware.bin");
    }

    return FirmwareFile(images);
}

FirmwareFile FirmwareFile::loadFromFile(const QString& filePath)
{
    QFileInfo fileInfo(filePath);

    if (fileInfo.isDir()) {
        return fromPlatformIOBuild(filePath);
    }

    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) {
        throw FirmwareLoadError(FirmwareLoadError::InvalidFile,
                                QString("Cannot open file: %1").arg(filePath));
    }

    QByteArray data = file.readAll();
    file.close();

    return FirmwareFile(filePath, data);
}

int FirmwareFile::totalSize() const
{
    int total = 0;
    for (const auto& image : m_images) {
        total += image.size();
    }
    return total;
}

QByteArray FirmwareFile::data() const
{
    // For backward compatibility, return the app firmware data
    for (const auto& image : m_images) {
        if (image.offset == 0x10000) {
            return image.data;
        }
    }
    return m_images.empty() ? QByteArray() : m_images[0].data;
}

QString FirmwareFile::fileName() const
{
    if (m_images.size() > 1) {
        return QString("%1 files").arg(m_images.size());
    }
    return m_images.empty() ? "No firmware" : m_images[0].fileName();
}

QString FirmwareFile::sizeDescription() const
{
    qint64 bytes = totalSize();
    QLocale locale;

    if (bytes < 1024) {
        return QString("%1 B").arg(bytes);
    } else if (bytes < 1024 * 1024) {
        return QString("%1 KB").arg(locale.toString(bytes / 1024.0, 'f', 1));
    } else {
        return QString("%1 MB").arg(locale.toString(bytes / (1024.0 * 1024.0), 'f', 2));
    }
}

bool FirmwareFile::isValid() const
{
    if (m_images.empty()) {
        return false;
    }
    return std::all_of(m_images.begin(), m_images.end(),
                       [](const FirmwareImage& img) { return img.isValid(); });
}

bool FirmwareFile::isComplete() const
{
    bool hasBootloader = false;
    bool hasPartitions = false;
    bool hasApp = false;

    for (const auto& image : m_images) {
        if (image.offset == 0x0000) hasBootloader = true;
        if (image.offset == 0x8000) hasPartitions = true;
        if (image.offset == 0x10000) hasApp = true;
    }

    return hasBootloader && hasPartitions && hasApp;
}

QString FirmwareFile::flashDescription() const
{
    QStringList parts;

    for (const auto& image : m_images) {
        QString name;
        switch (image.offset) {
        case 0x0000: name = "bootloader"; break;
        case 0x8000: name = "partitions"; break;
        case 0x10000: name = "app"; break;
        default: name = image.fileName(); break;
        }

        FirmwareFile tempFile({image});
        QString desc = QString("%1 @ 0x%2 (%3)")
                           .arg(name)
                           .arg(image.offset, 0, 16)
                           .arg(tempFile.sizeDescription());
        parts.append(desc);
    }

    return parts.join(", ");
}

File: windows/Models/FirmwareFile.cs
// FAME Smart Flasher - C# Port
// Copyright 2025 Fyrby Additive Manufacturing & Engineering
// SPDX-License-Identifier: Proprietary

using System.IO;

namespace FAMESmartFlasher.Models;

/// <summary>
/// Represents a single firmware image with its flash offset
/// </summary>
public class FirmwareImage
{
    public required string FilePath { get; init; }
    public required byte[] Data { get; init; }
    public required uint Offset { get; init; }

    public int Size => Data.Length;
    public string FileName => Path.GetFileName(FilePath);

    /// <summary>
    /// Check if the file appears to be valid ESP32 firmware
    /// ESP32 firmware magic byte is 0xE9
    /// </summary>
    public bool IsValid => Data.Length >= 8 && Data[0] == 0xE9;
}

/// <summary>
/// Represents a complete firmware package (bootloader, partitions, app)
/// ESP32-C3 flash layout:
/// - 0x0000: bootloader.bin (second-stage bootloader)
/// - 0x8000: partitions.bin (partition table)
/// - 0x10000: firmware.bin (application)
/// </summary>
public class FirmwareFile
{
    public List<FirmwareImage> Images { get; init; } = new();

    /// <summary>
    /// Single-file constructor
    /// Detects merged firmware (starting at 0x0) vs app-only (at 0x10000)
    /// </summary>
    public FirmwareFile(string filePath, byte[] data)
    {
        var fileName = Path.GetFileName(filePath).ToLowerInvariant();
        var isMergedBinary = fileName.Contains("merged") ||
                            fileName.Contains("factory") ||
                            fileName.Contains("combined") ||
                            fileName.Contains("full");

        var offset = isMergedBinary ? 0x0000u : 0x10000u;
        Images = new List<FirmwareImage>
        {
            new FirmwareImage
            {
                FilePath = filePath,
                Data = data,
                Offset = offset
            }
        };
    }

    /// <summary>
    /// Multi-file constructor for complete firmware package
    /// </summary>
    public FirmwareFile(List<FirmwareImage> images)
    {
        Images = images.OrderBy(i => i.Offset).ToList();
    }

    /// <summary>
    /// Create from PlatformIO build directory
    /// </summary>
    public static FirmwareFile FromPlatformIOBuild(string directoryPath)
    {
        var images = new List<FirmwareImage>();

        var fileOffsets = new (string FileName, uint Offset)[]
        {
            ("bootloader.bin", 0x0000),
            ("partitions.bin", 0x8000),
            ("firmware.bin", 0x10000)
        };

        foreach (var (fileName, offset) in fileOffsets)
        {
            var filePath = Path.Combine(directoryPath, fileName);
            if (File.Exists(filePath))
            {
                var data = File.ReadAllBytes(filePath);
                images.Add(new FirmwareImage
                {
                    FilePath = filePath,
                    Data = data,
                    Offset = offset
                });
            }
        }

        if (images.Count == 0)
            throw new FirmwareLoadException("No firmware files found in directory");

        // At minimum we need firmware.bin
        if (!images.Any(i => i.Offset == 0x10000))
            throw new FirmwareLoadException("Missing firmware.bin");

        return new FirmwareFile(images);
    }

    public int TotalSize => Images.Sum(i => i.Size);
    public int Size => TotalSize;

    public byte[] Data =>
        // For backward compatibility, return the app firmware data
        Images.FirstOrDefault(i => i.Offset == 0x10000)?.Data ?? Array.Empty<byte>();

    public string FileName =>
        Images.Count > 1 ? $"{Images.Count} files" : Images.FirstOrDefault()?.FileName ?? "No firmware";

    public string SizeDescription
    {
        get
        {
            var kb = TotalSize / 1024.0;
            return kb >= 1024
                ? $"{kb / 1024:F1} MB"
                : $"{kb:F1} KB";
        }
    }

    /// <summary>
    /// Check if the firmware package is valid
    /// For app firmware at 0x10000 or merged firmware at 0x0, check ESP32 magic byte (0xE9)
    /// Bootloader and partition table have different formats
    /// </summary>
    public bool IsValid => Images.Any(i => (i.Offset == 0x10000 || i.Offset == 0x0) && i.IsValid);

    /// <summary>
    /// Check if this is a complete package (has bootloader, partitions, and app)
    /// </summary>
    public bool IsComplete
    {
        get
        {
            var offsets = Images.Select(i => i.Offset).ToHashSet();
            return offsets.Contains(0x0000) && offsets.Contains(0x8000) && offsets.Contains(0x10000);
        }
    }

    /// <summary>
    /// Description of what will be flashed
    /// </summary>
    public string FlashDescription
    {
        get
        {
            var parts = Images.Select(image =>
            {
                var name = image.Offset switch
                {
                    0x0000 => "bootloader",
                    0x8000 => "partitions",
                    0x10000 => "app",
                    _ => image.FileName
                };

                var kb = image.Size / 1024.0;
                var size = kb >= 1024 ? $"{kb / 1024:F1} MB" : $"{kb:F1} KB";
                return $"{name} @ 0x{image.Offset:X} ({size})";
            });

            return string.Join(", ", parts);
        }
    }
}

/// <summary>
/// Firmware load errors
/// </summary>
public class FirmwareLoadException : Exception
{
    public FirmwareLoadException(string message) : base(message) { }
}
